### ISSUE TITLE
Can't reconnect until invalid transaction is rolled back

### ISSUE THREAD

**Issue by tiredpixel** [tiredpixel/pikka-bird-server-py#1]
After the server was running for some time, it stopped handling requests, reporting `Can't reconnect until invalid transaction is rolled back`. This appears to be coming from the database connection; possibly, Pikka Bird's db connection isn't set up correctly for teardown.

After restarting the server, clients were able to send requests again, without needing adjustment.

Possibly related to
https://stackoverflow.com/questions/23301968/invalid-transaction-persisting-across-requests

_Server_

```
[2015-05-28 19:05:52,647] ERROR -- pikka_bird_server: Exception on /collections [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python2.7/dist-packages/pikka_bird_server/routes/collections.py", line 28, in create
    machine = Machine.find(address=request.remote_addr)
  File "/usr/local/lib/python2.7/dist-packages/pikka_bird_server/models/base.py", line 34, in find
    r = cls.query.filter_by(**kwargs).first()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2444, in first
    ret = list(self[0:1])
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2280, in __getitem__
    return list(res)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2515, in __iter__
    return self._execute_and_instances(context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2530, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1078, in _execute_context
    None, None)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1332, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1071, in _execute_context
    conn = self._revalidate_connection()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 391, in _revalidate_connection
    "Can't reconnect until invalid "
StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back [SQL: u'SELECT machines.id AS machines_id, machines.created_at AS machines_created_at, machines.updated_at AS machines_updated_at, machines.address AS machines_address, machines.hostname AS machines_hostname \nFROM machines \nWHERE machines.address = %(address_1)s \n LIMIT %(param_1)s'] [parameters: [immutabledict({})]]
```

_Client_

```
[2015-05-28 19:13:57,286] INFO -- 14280: COLLECTING
[2015-05-28 19:13:57,286] INFO -- 14280: COLLECTING System
[2015-05-28 19:13:57,387] INFO -- 14280: COLLECTED (0 s)
[2015-05-28 19:13:57,387] INFO -- 14280: SENDING http://example.tiredpixel.com:8042/collections binary (1493 b)
[2015-05-28 19:13:57,395] ERROR -- 14280: SENT 500 {
  "message": "(sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back [SQL: u'SELECT machines.id AS machines_id, machines.created_at AS machines_created_at, machines.updated_at AS machines_updated_at, machines.address AS machines_address, machines.hostname AS machines_hostname \\nFROM machines \\nWHERE machines.address = %(address_1)s \\n LIMIT %(param_1)s'] [parameters: [immutabledict({})]]"
} (0 s)
[2015-05-28 19:13:57,395] INFO -- 14280: WATCHING Perfectly Normal Beasts (17 s)
```

**Comment by tiredpixel** [comment]
I hadn't seen this since I filed in May, but it's happened again, with a very similar stacktrace.

**Comment by Tjorriemorrie** [comment]
Happening to me too; restarting the server resolved it. Is this library stable?

**Comment by tiredpixel** [comment]
@Tjorriemorrie, thanks for letting me know. I suspect I've made a mistake in the database integration using Flask, causing cleanup to not happen correctly on failure. The stacktrace I saw before makes me think the connection then gets stuck in that invalid state without auto-rollback. I'd be grateful for additional stacktraces, if you've been experiencing the problem, too, even if just to confirm things are happening in the same place.

No, this library is not yet stable; the latest release was 0.1.0, with not even the payload structure being stabilised, and note the README:

> Pikka Bird is currently in a draft phase, which means that payloads and schemas might be changed in a backwards-incompatible fashion. Although it is unlikely, in extreme cases this could require you to reinstall with an empty database. If this upsets you too much, please wave and come back later. :) Currently, it is not recommended that you use Pikka Bird as a replacement for any of your usual monitoring tools.

I've used it to assist with other tools in production, where it's often worked for weeks at a time without issue, but it's by no means production-ready, and shouldn't replace any of your other monitoring. So, it should be considered preview only at this stage.
